sycl/include/sycl/ext/oneapi/owner_less.hpp

```diff
@@ -0,0 +1,101 @@
+
+#pragma once
+
+#include <sycl/accessor.hpp>
+#include <sycl/buffer.hpp>
+#include <sycl/context.hpp>
+#include <sycl/detail/defines_elementary.hpp>
+#include <sycl/device.hpp>
+#include <sycl/event.hpp>
+#include <sycl/ext/oneapi/weak_object.hpp>
+#include <sycl/kernel.hpp>
+#include <sycl/kernel_bundle.hpp>
+#include <sycl/platform.hpp>
+#include <sycl/queue.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+
+namespace detail {
+template <typename SyclObject> struct owner_less_base {
+  bool operator()(const SyclObject &lhs, const SyclObject &rhs) const noexcept {
+    return lhs.ext_oneapi_owner_before(rhs);
+  }
+
+  bool operator()(const weak_object<SyclObject> &lhs,
+                  const weak_object<SyclObject> &rhs) const noexcept {
+    return lhs.owner_before(rhs);
+  }
+
+  bool operator()(const SyclObject &lhs,
+                  const weak_object<SyclObject> &rhs) const noexcept {
+    return lhs.ext_oneapi_owner_before(rhs);
+  }
+
+  bool operator()(const weak_object<SyclObject> &lhs,
+                  const SyclObject &rhs) const noexcept {
+    return lhs.owner_before(rhs);
+  }
+};
+} // namespace detail
+
+template <typename SyclObject> struct owner_less;
+
+template <>
+struct owner_less<context> : public detail::owner_less_base<context> {};
+template <>
+struct owner_less<device> : public detail::owner_less_base<device> {};
+template <> struct owner_less<event> : public detail::owner_less_base<event> {};
+template <>
+struct owner_less<kernel> : public detail::owner_less_base<kernel> {};
+template <>
+struct owner_less<kernel_id> : public detail::owner_less_base<kernel_id> {};
+template <>
+struct owner_less<platform> : public detail::owner_less_base<platform> {};
+template <> struct owner_less<queue> : public detail::owner_less_base<queue> {};
+
+template <bundle_state State>
+struct owner_less<device_image<State>>
+    : public detail::owner_less_base<device_image<State>> {};
+
+template <bundle_state State>
+struct owner_less<kernel_bundle<State>>
+    : public detail::owner_less_base<kernel_bundle<State>> {};
+
+template <typename DataT, int Dimensions, typename AllocatorT>
+struct owner_less<buffer<DataT, Dimensions, AllocatorT>>
+    : public detail::owner_less_base<buffer<DataT, Dimensions, AllocatorT>> {};
+
+template <typename DataT, int Dimensions, access_mode AccessMode,
+          target AccessTarget, access::placeholder isPlaceholder>
+struct owner_less<
+    accessor<DataT, Dimensions, AccessMode, AccessTarget, isPlaceholder>>
+    : public detail::owner_less_base<accessor<DataT, Dimensions, AccessMode,
+                                              AccessTarget, isPlaceholder>> {};
+
+template <typename DataT, int Dimensions, access_mode AccessMode>
+struct owner_less<host_accessor<DataT, Dimensions, AccessMode>>
+    : public detail::owner_less_base<
+          host_accessor<DataT, Dimensions, AccessMode>> {};
```

Comment:
nit, feel free to ignore.

I'd rather see something like 

```
namespace detail {
template <typename T>
struct owner_less_unsupport {
static_assert(always_false_v<T>);
}
}

template <typename T>
struct owner_less : public std::conditional<....,
                                            detail::owner_less_base<T>,
                                            detail::owner_less_unsupported<T>> {};
```

using some utilities like https://github.com/intel/llvm/blob/ff59e4185b0c759fd14a369c8808d903faa2d2d5/sycl/include/sycl/properties/accessor_properties.hpp#L106-L115 for the condition.

Reply:
So instead of having a handful specializations we would have a generic class with checks for all the different classes we should support? Since we would still need to add new `is_X` traits for new classes and add them here, I am not sure I see the benefit to that approach.

Reply:
I have a feeling that I see the same pattern in multiple places. As such, new `is_X` cost would be spread over multiple places (maybe not now, but soon enough). For example, `is_*access*` already has other usage - the link above.

Reply:
I suppose for the ones with multiple template arguments this approach would make it cleaner, but I am reluctant to add these helpers as part of this PR as it is already fairly big. I will try and do some refactoring after this is merged.